### ISSUE TITLE
adding new "handler" in multiprocessing context 

### ISSUE THREAD

**Issue by danielhrisca** [Delgan/loguru#421]
I'm trying to add a new handler after creating child processes. The flow is like this:

1. configure loguru with a file handler in the parent process (enqueue = True)
2. start the child process
3. after some time add a new file handler in the parent process (enqueue = True)

What I hoped for is to have the messages from the parent and child processes in the seconds file, but only the parent messages are stored

**Comment by Delgan** [comment]
If the handler is added after the child process is started, that means the child process won't inherit the handler and therefore the messages won't be logged. There is some additional explanations in the documentation: [Compatibility with `multiprocessing` using `enqueue` argument](https://loguru.readthedocs.io/en/stable/resources/recipes.html#compatibility-with-multiprocessing-using-enqueue-argument).

One possible workaround is to add both handlers before spawning the child process, but use a `filter` to dynamically enable and disable the second one.

```python
class LogFilter:
    def __init__(self, enabled=False):
        self.enabled = enabled

    def filter(self, record):
        return self.enabled

log_filter = LogFilter()
logger.add("file_1.log", enqueue=True)
logger.add("file_2.log", enqueue=True, filter=log_filter.filter)

# Spawn child process and log messages

log_filter.enabled = True

# Log messages from child process now appear in 2nd handler
```

**Comment by danielhrisca** [comment]
Hey @Delgan, 
thank you for the information.

The `enable` approach does not work for my use case:
- the parent process has a "general" file log enabled from the start, and this one gets messages from the child as well
- the child process further spawns other processes that also get the initial logger configuration and can send messages to the "general" file log
- the child process executed a number of tasks, and for each of them I would like to get a separate file log. This one would be like a snapshot of the "general" log but only holding the messages vehiculated while the task is executed

Do you think such an use case if possible?

**Comment by Delgan** [comment]
You'll probably need to explicitly pass the `logger` object to your child process during initialization (see link to the documentation I posted previously).

Something like that for example:

```python
def worker(logger, num):
     logger.info("Logging from worker child process #{}", num)

def worker_spawner(num):
    handler_id = logger.add(f"worker_{num}.log", enqueue=True)
    process = multiprocessing.Process(target=worker, args=(logger, num))
    process.start()
    process.join()
    logger.remove(handler_id)
```

**Comment by danielhrisca** [comment]
The first process is the GUI process; the worker process is started once and it usually handles all the tasks until the application is closed. The other processes started by the worker process can also be alive for several tasks.

**Comment by Delgan** [comment]
What is your concern with the proposed workaround of passing the `logger` via `args`?

This seems to me to be the only solution to handle multiple file sinks for multiple processes. It's not possible to add an handler after starting the child process, because at that point, the new process evolves independently. It is initialized by copying the state of the parent process, but it does not have easy access to its memory afterward.

**Comment by danielhrisca** [comment]
> What is your concern with the proposed workaround of passing the `logger` via `args`?

This works for the first log ``worker_0.log``. Once Task 0 is finished then the log file is saved. The problem is that when the next task is started we still have the same GUI and worker processes from the beginning because the worker process is not spawned again for the new task. The application is running on Windows and starting new processes is expensive.

**Comment by Delgan** [comment]
Sorry I don't understand your problem. Why not `add()` and `remove()` a file handler for each task then? As long as the `logger` is properly passed to any child process you spawn, it should work fine. Maybe sharing a minimal reproducible example of your issue would help me to suggest a suitable solution.

**Comment by danielhrisca** [comment]
Here is very simplified example

```python
from loguru import logger
import os
import multiprocessing
from time import sleep, perf_counter


def worker(logger=None):
    procs = []
    for i in range(3):
        procs.append(multiprocessing.Process(target=slave, args=(logger,)))

    for p in procs:
        p.start()

    t = perf_counter()
    while perf_counter() - t < 50:
        logger.info("Worker")
        sleep(0.5)

    for p in procs:
        p.join()


def slave(logger):
    t = perf_counter()
    pid = os.getpid()
    while perf_counter() - t < 50:
        logger.info(f"Slave {pid}")
        sleep(0.5)


def main(logger):
    logger.add(
        "general.log",
        enqueue=True,
        level="DEBUG",
        format="{time} {name} {line} - {message}",
        encoding="utf8",
    )

    myworker = multiprocessing.Process(target=worker, args=(logger,))
    myworker.start()

    for i in range(20):
        logger.info("MAIN")
        sleep(1)

    logger.add(
        "task.log",
        enqueue=True,
        level="DEBUG",
        format="{time} {name} {line} - {message}",
        encoding="utf8",
    )

    for i in range(20):
        logger.info("MAIN")
        sleep(1)

    myworker.join()


if __name__ == "__main__":
    multiprocessing.freeze_support()
    logger.remove()
    main(logger)

```

and here are the logs generated 
[general.log](https://github.com/Delgan/loguru/files/6252349/general.log)
[task.log](https://github.com/Delgan/loguru/files/6252350/task.log)

**Comment by Delgan** [comment]
Thanks for the example!

I was thinking of adding the handlers from within the child process but doing so means the `tasks.log` won't contain logs from the parent. Also, adding the handler after starting the process means it's not copied and consequently not used by the child process. 

So it's more complicated than expected. The handler needs to be added in the parent for sure. I think you may need to use a custom sink and manage log files by yourself. Here is a proof of concept:

```python
from loguru import logger
import os
import multiprocessing
from time import sleep, perf_counter


class MultiFilesSink:
    def __init__(self):
        self._files = {}

    def write(self, message):
        # TODO: Remove no longer used file objects
        task = message.record["extra"].get("task", None)
        if task is None:
            for file in self._files.values():
                file.write(message)
        else:
            if task not in self._files:
                self._files[task] = open(f"task_{task}.log", "w")
            self._files[task].write(message)


def worker(logger=None):
    procs = []
    for i in range(3):
        task_logger = logger.bind(task=i)  # You could also use 'record["process"].id' instead
        procs.append(multiprocessing.Process(target=slave, args=(task_logger,)))

    for p in procs:
        p.start()

    t = perf_counter()
    while perf_counter() - t < 50:
        logger.info("Worker")
        sleep(0.5)

    for p in procs:
        p.join()


def slave(logger):
    t = perf_counter()
    pid = os.getpid()
    while perf_counter() - t < 50:
        logger.info(f"Slave {pid}")
        sleep(0.5)


def main(logger):
    logger.add(
        "general.log",
        enqueue=True,
        level="DEBUG",
        format="{time} {name} {line} - {message}",
        encoding="utf8",
    )
    logger.add(
        MultiFilesSink(),
        enqueue=True,
        level="DEBUG",
        format="{time} {name} {line} - {message}",
    )

    myworker = multiprocessing.Process(target=worker, args=(logger,))
    myworker.start()

    for i in range(20):
        logger.info("MAIN")
        sleep(1)

    for i in range(20):
        logger.info("MAIN")
        sleep(1)

    myworker.join()


if __name__ == "__main__":
    multiprocessing.freeze_support()
    logger.remove()
    main(logger)
```

In the end, this looks quite convoluted. I would suggest using a single file, but separating the logs for each tasks during a post-processing step instead (possibly by passing a custom function to the `rotation` parameter).

**Comment by danielhrisca** [comment]
I'll try this, but now I have another problem:
* `main` start the `worker` and passed the logger to it
* `worker` starts the `slaves` and passes the logger to them as well
* at some point the `worker` and `slaves` are re-created:
    * `main` send the termination request
    * the `slaves` are stopped gracefully and they call `logger.complete()`
    * `worker` stops gracefully and calls `logger.complete()`
    * `main` starts the `worker` again , and the `worker` starts the `slaves`
* after a few logging messages are send from the  `slaves` and `worker` the next log attempt halts indefinitely

Did I miss something about terminating a process that received a logger handle?

**Comment by Delgan** [comment]
It should work fine. Are you using the `signal` module by any chance? The logging methods are not re-entrant. If the call is "paused" or if the `logger` is re-used from within the handler, it may hang forever because the internal lock will never be released.
